src/app/qgsrelationmanagerdialog.cpp

```diff
@@ -15,93 +15,184 @@
+        return QStyledItemDelegate::createEditor( parentWidget, option, index );
+
+      return nullptr;
+    }
+  private:
+    QList<int> mEditableColumns;
+};
+#endif
+
 QgsRelationManagerDialog::QgsRelationManagerDialog( QgsRelationManager *relationMgr, QWidget *parent )
   : QWidget( parent )
   , Ui::QgsRelationManagerDialogBase()
   , mRelationManager( relationMgr )
 {
   setupUi( this );
+
+  mRelationsTree->header()->setSectionResizeMode( QHeaderView::ResizeToContents );
+  mRelationsTree->setItemDelegate( new RelationNameEditorDelegate( QList<int>( {0} ), mRelationsTree ) );
+
   connect( mBtnAddRelation, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnAddRelation_clicked );
+  connect( mActionAddPolymorphicRelation, &QAction::triggered, this, &QgsRelationManagerDialog::mActionAddPolymorphicRelation_triggered );
+  connect( mActionEditPolymorphicRelation, &QAction::triggered, this, &QgsRelationManagerDialog::mActionEditPolymorphicRelation_triggered );
   connect( mBtnDiscoverRelations, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnDiscoverRelations_clicked );
   connect( mBtnRemoveRelation, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnRemoveRelation_clicked );
 
   mBtnRemoveRelation->setEnabled( false );
-  connect( mRelationsTable->selectionModel(), &QItemSelectionModel::selectionChanged, this, &QgsRelationManagerDialog::onSelectionChanged );
+  mBtnAddRelation->setPopupMode( QToolButton::MenuButtonPopup );
+  mBtnAddRelation->addAction( mActionAddPolymorphicRelation );
+  mBtnAddRelation->addAction( mActionEditPolymorphicRelation );
+
+  connect( mRelationsTree->selectionModel(), &QItemSelectionModel::selectionChanged, this, &QgsRelationManagerDialog::onSelectionChanged );
 }
 
 void QgsRelationManagerDialog::setLayers( const QList< QgsVectorLayer * > &layers )
 {
   mLayers = layers;
 
   const QList<QgsRelation> &relations = mRelationManager->relations().values();
-
-  const auto constRelations = relations;
-  for ( const QgsRelation &rel : constRelations )
+  for ( const QgsRelation &rel : relations )
   {
+    // the generated relations for polymorphic relations should be ignored,
+    // they are generated when the polymorphic relation is added to the table
+    if ( rel.type() == QgsRelation::Generated )
+      continue;
+
     addRelation( rel );
   }
+  const QList<QgsPolymorphicRelation> &polymorphicRelations = mRelationManager->polymorphicRelations().values();
+  for ( const QgsPolymorphicRelation &polymorphicRel : polymorphicRelations )
+  {
+    addPolymorphicRelation( polymorphicRel );
+  }
 
-  mRelationsTable->sortByColumn( 0, Qt::AscendingOrder );
+  mRelationsTree->sortByColumn( 0, Qt::AscendingOrder );
 }
 
-void QgsRelationManagerDialog::addRelation( const QgsRelation &rel )
+bool QgsRelationManagerDialog::addRelation( const QgsRelation &rel )
 {
-  if ( ! rel.isValid() )
-    return;
+  return addRelationPrivate( rel );
```

Comment:
this one should have kept the testing for generated relation, no? it is a public method, and it should handle all relations I think.

Reply:
There is a check in the private method anyway, that's why I removed it

Reply:
It should not handle all the relations, the generated relations should only be added by calling `addPolymorphicRelation`.

Reply:
Well, it should handle all the relations but the generated ones, probably makes sense to be private. I will change that.